src/core_plugins/kibana/public/dashboard/dashboard.js

```diff
@@ -52,125 +52,87 @@ uiRoutes
+      $scope.getBrushEvent = () => brushEvent(dashboardState.getAppState());
+      $scope.getFilterBarClickHandler = () => filterBarClickHandler(dashboardState.getAppState());
+      $scope.hasExpandedPanel = () => $scope.expandedPanel !== null;
+      $scope.getDashTitle = () => {
+        return dashboardState.dashboard.lastSavedTitle || `${dashboardState.dashboard.title} (unsaved)`;
+      };
+      $scope.newDashboard = () => { kbnUrl.change(DashboardConstants.CREATE_NEW_DASHBOARD_URL, {}); };
+      $scope.saveState = () => dashboardState.saveState();
 
-        if (dash.id) {
-          docTitle.change(dash.title);
+      $scope.toggleExpandPanel = (panelIndex) => {
+        if ($scope.expandedPanel && $scope.expandedPanel.panelIndex === panelIndex) {
+          $scope.expandedPanel = null;
+        } else {
+          $scope.expandedPanel =
+            dashboardState.getPanels().find((panel) => panel.panelIndex === panelIndex);
         }
+      };
 
-        initPanelIndexes();
-
-        // watch for state changes and update the appStatus.dirty value
-        stateMonitor = stateMonitorFactory.create($state, stateDefaults);
-        stateMonitor.onChange((status) => {
-          $appStatus.dirty = status.dirty;
-        });
-
-        $scope.$on('$destroy', () => {
-          stateMonitor.destroy();
-          dash.destroy();
-
-          // Remove dark theme to keep it from affecting the appearance of other apps.
-          setDarkTheme(false);
-        });
-
-        $scope.$emit('application.load');
-      }
-
-      function initPanelIndexes() {
-        // find the largest panelIndex in all the panels
-        let maxIndex = getMaxPanelIndex();
-
-        // ensure that all panels have a panelIndex
-        $scope.state.panels.forEach(function (panel) {
-          if (!panel.panelIndex) {
-            panel.panelIndex = maxIndex++;
-          }
-        });
-      }
-
-      function getMaxPanelIndex() {
-        let maxId = $scope.state.panels.reduce(function (id, panel) {
-          return Math.max(id, panel.panelIndex || id);
-        }, 0);
-        return ++maxId;
-      }
+      $scope.filterResults = function () {
+        dashboardState.setQuery($scope.model.query);
+        dashboardState.updateFilters(queryFilter);
+        $scope.refresh();
+      };
 
-      function updateQueryOnRootSource() {
-        const filters = queryFilter.getFilters();
-        if ($state.query) {
-          dash.searchSource.set('filter', _.union(filters, [{
-            query: $state.query
-          }]));
-        } else {
-          dash.searchSource.set('filter', filters);
-        }
-      }
+      // called by the saved-object-finder when a user clicks a vis
+      $scope.addVis = function (hit) {
+        pendingVisCount++;
+        const maxPanelIndex = PanelUtils.getMaxPanelIndex(dashboardState.getPanels());
+        dashboardState.getPanels().push(createPanelState(hit.id, 'visualization', maxPanelIndex));
```

Comment:
I find the fact that `getPanels` is used to access the panels for manipulation quite counter-intuitive. The presence of a getter suggests that it is either read-only or there is a separate setter to manipulate the value (like for `getQuery`/`setQuery` of `DashboardState`).